src/poetry/packages/locker.py

```diff
@@ -420,6 +421,10 @@ def set_lock_data(self, root: Package, packages: list[Package]) -> bool:
         return False
 
     def _write_lock_data(self, data: TOMLDocument) -> None:
+        if self._contains_credential:
+            logger.warning(
+                "A package URL contains credentials and will be written to poetry.lock."
```

Comment:
This looks like a frustrating error message - I might plausibly have hundreds of packages in `poetry.lock`, being told that _one_ of them has credentials is going to leave me with work to do...

Can this warning say which package is the problematic one?

Reply:
The warning is not triggered by a package. It is displayed if credentials are in the URL of an added source in `package.source` section.

Reply:
Those credentials are associated with a particular package, no?  eg in your testcase, it's package `A` that triggers the warning.

If I have a lockfile with hundreds of packages in it, then a more helpful warning would point me at `A` rather than only telling me that there's a problem, somewhere.

Reply:
When I got this issue, the URL coming from the `package.source` was added to every packages in the lock file. So it's not useful. However, adding the URL in the message could help the user:

```
Package URL contain credentials (https://alice:a_passwd@pypi.domain.tld/site/prod/+simple) and will be written to poetry.lock.
```

From file:
```
[package.source]
type = "legacy"
url = "https://alice:a_passwd@pypi.domain.tld/site/prod/+simple"
reference = "whatever"
```

Reply:
This LGTM, except I think we should resolve this thread. Can we include the URL?

Reply:
@neersighted I can modify `_contains_credential` to be the URL instead of a boolean status. In such case, the name fits less well. If it's still ok fo you, I'll be happy to modify it. An other way would be to rename it but it will also require to rename it in the poetry-plugin-export  repository. The third way is to not add the URL. My favorite way is the first one (modifying the content of `_contains_credential`). What do you prefer?